psutil/arch/windows/process_info.c

```diff
@@ -804,10 +870,72 @@ psutil_get_cmdline(long pid) {
     PyObject *py_unicode = NULL;
     LPWSTR *szArglist = NULL;
     int nArgs, i;
+    HANDLE hProcess;
+    int windows_version;
+    ULONG ret_length = 4096;
+    NTSTATUS status;
+    char * cmdline_buffer = NULL;
+    WCHAR * cmdline_buffer_wchar = NULL;
+    PUNICODE_STRING tmp = NULL;
+    DWORD string_size;
+    _NtQueryInformationProcess NtQueryInformationProcess = NULL;
+
+    windows_version = get_windows_version();
+    if (windows_version >= WINDOWS_81) {
+        // by defaut, still use PEB (if command line params have been patched in PEB, we will get the actual ones)
+        if (psutil_get_process_data(pid, KIND_CMDLINE, &data, &size) != 0) {
+            PyErr_Clear(); // reset that we had an error, and retry with NtQueryInformationProcess
+            // if it fails, fallback to NtQueryInformationProcess (for protected processes)
+            NtQueryInformationProcess = get_NtQueryInformationProcess();
+            if (NtQueryInformationProcess == NULL) {
+                PyErr_SetFromWindowsErr(0);
+                goto out;
+            }
 
-    if (psutil_get_process_data(pid, KIND_CMDLINE, &data, &size) != 0)
-        goto out;
+            cmdline_buffer = calloc(4096, 1);
+            if (cmdline_buffer == NULL) {
+                PyErr_NoMemory();
+                goto out;
+            }
+
+            hProcess = psutil_handle_from_pid(pid, PROCESS_QUERY_LIMITED_INFORMATION);
+            if (hProcess == NULL) {
+                PyErr_SetFromWindowsErr(0);
+                goto out;
+            }
+            status = NtQueryInformationProcess(
+                hProcess,
+                60, // ProcessCommandLineInformation
+                cmdline_buffer,
+                ret_length,
+                &ret_length
+            );
+            CloseHandle(hProcess);
+            if (!NT_SUCCESS(status)) {
+                PyErr_SetFromWindowsErr(0);
+                goto out;
+            }
```

Comment:
OK this is a bit subtle. =) `CloseHandle` may also fail and set errno, so in order to not overwrite the original error code you want to call it *after* you do `PyErr_SetFromWindowsErr(0);`. So in summary I think what you want to do this:

```c
            if (!NT_SUCCESS(status)) {
                PyErr_SetFromWindowsErr(0);
                CloseHandle(hProcess);
                goto out;
            }

            CloseHandle(hProcess);
  ```